### PR TITLE
Strengthen implementation of MetadataHelpers.GetInfoForImmediateNamespaceMembers against unconventional/obfuscated namespace names in metadata.

### LINKED ISSUE #7396: [VS Crash] CSharpAddImportCodeFixProvider encountered an error and has been disabled.
I get this warning at the top of the screen in VS 2015 Pro (RTM, and now also Update 1):

'CSharpAddImportCodeFixProvider' encountered an error and has been disabled.

This is a frequent harbinger of an imminent IDE crash.  However, not all IDE crashes are preceded by these warnings. 

I'm usually typing new code and a dot that would normally show an intellisense pop-up will cause this.  Also happens when pasting code.  Also happens when adjusting curly braces.

My solution has bad namespace hygiene.  I think that's related to the problem, and can make this difficult to repro outside of my solution.

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #7396.

@dotnet/roslyn-compiler Please review.

### CODE REVIEW

File: src/Compilers/Core/Portable/MetadataReader/MetadataHelpers.cs [hunk 1]
```diff
@@ -790,6 +804,37 @@ internal static string UnmangleMetadataNameForArity(string emittedTypeName, int
             } // using
 
             types = nestedTypes;
+
+            // Merge pairs with the same key
+            if (possiblyHavePairsWithDuplicateKey)
+            {
+                var names = new Dictionary<string, int>(nameComparer);
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/MetadataReader/MetadataHelpers.cs]: Does it make sense to set the capacity here to `nestedNamespaces.Count` given that is at the least an upper bound on the number of entries.

File: src/Compilers/Core/Portable/MetadataReader/MetadataHelpers.cs [hunk 2]
```diff
@@ -737,7 +743,7 @@ internal static string UnmangleMetadataNameForArity(string emittedTypeName, int
                     }
 
                     // Account for the dot following THIS namespace name.
-                    if (namespaceNameLength != 0)
+                    if (!isGlobalNamespace)
                     {
                         namespaceNameLength++;
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/MetadataReader/MetadataHelpers.cs]: Why do we increment this on every iteration of the loop?  Given that it seems to track the value passed into the function it seems that it should only be incremented once?

I'm positive that I'm just missing something basic here.

Comment (AlekseyTs) [inline src/Compilers/Core/Portable/MetadataReader/MetadataHelpers.cs]: What loop? :-)

File: src/Compilers/Core/Portable/MetadataReader/MetadataHelpers.cs (hunk 1 above)
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/MetadataReader/MetadataHelpers.cs]: Sure, I'll do that.

**Comment by cston** [comment]
LGTM

File: src/Compilers/Core/Portable/MetadataReader/MetadataHelpers.cs (hunk 2 above)
Comment (jaredpar) [inline src/Compilers/Core/Portable/MetadataReader/MetadataHelpers.cs]: I can't read code :(

File: src/Compilers/Core/Portable/MetadataReader/MetadataHelpers.cs [hunk 3]
```diff
@@ -756,6 +763,12 @@ internal static string UnmangleMetadataNameForArity(string emittedTypeName, int
                         else
                         {
                             // This is a new child namespace
+                            if (cmp > 0)
+                            {
+                                // The sort order is violated for child namespace names. Obfuscation is the likely reason for this.
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/MetadataReader/MetadataHelpers.cs]: Where is this sort order documented?  It seems like we are relying here on the ordering of `StringComparer` to match up with that of metadata ordering.  Is that guaranteed to be the same?

Comment (AlekseyTs) [inline src/Compilers/Core/Portable/MetadataReader/MetadataHelpers.cs]: The doc comment for typesByNS documents expected sort order for groups. For conventional namespace names, nested namespace names will be in the same sort order.

**Comment by jaredpar** [comment]
:+1:

**Pull request merged**